classes/booking_manager.php

```diff
@@ -414,25 +471,37 @@ public function process() {
                         $session,
                         $this->facetoface,
                         $this->course,
-                        $entry->discountcode,
-                        $this->transform_notification_type($entry->notificationtype),
+                        $entry->discountcode ?? '',
+                        $this->transform_notification_type($entry->notificationtype ?? ''),
                         $statuscode,
                         $user->id,
                         !$this->suppressemail,
                     );
-
-                    continue;
                 }
 
                 // Handle attendance.
-                if (
-                    in_array($statuscode, [
-                    MDL_F2F_STATUS_NO_SHOW,
-                    MDL_F2F_STATUS_PARTIALLY_ATTENDED,
-                    MDL_F2F_STATUS_FULLY_ATTENDED,
-                    ])
-                ) {
+                if ($isattendance) {
+                    // If booking into attendance but the user hasn't been signed up yet (e.g. directly marking as attended),
+                    // then sign the user up.
+                    $alreadysignedup = $DB->record_exists('facetoface_signups', ['sessionid' => $session->id, 'userid' => $user->id]);
+                    if (!$alreadysignedup) {
+                        // We use booked/waitlisted for making the signup,
+                        // and then the actual status from the CSV when taking attendance.
+                        $signupstatus = $session->datetimeknown ? MDL_F2F_STATUS_BOOKED : MDL_F2F_STATUS_WAITLISTED;
```

Comment:
I'm guessing if the datetime is known, waitlisted is discarded as a valid default option, because we're filling the entry to ensure we can take attendance further down (i.e. it's not important what the value is)